files/startup-script-stdlib.sh

```diff
@@ -0,0 +1,230 @@
+
+  # Color variables
+  if [[ -n "${COLOR:-}" ]]; then
+    readonly NC='\033[0m'        # no color
+    readonly RED='\033[0;31m'    # error
+    readonly GREEN='\033[0;32m'  # info
+    readonly BLUE='\033[0;34m'   # debug
+  else
+    readonly NC=''
+    readonly RED=''
+    readonly GREEN=''
+    readonly BLUE=''
+  fi
+
+  return 0
+}
+
+stdlib::init_directories() {
+  if ! [[ -e "${VARDIR}" ]]; then
+    install -d -m 0755 -o 0 -g 0 "${VARDIR}"
+  fi
+}
+
+##
+# Get a metadata key.  When used without -o, this function is guaranteed to
+# produce no output on STDOUT other than the retrieved value.  This is intended
+# to support the use case of
+# FOO="$(stdlib::metadata_get -k instance/attributes/foo)"
+#
+# If the requested key does not exist, the error code will be 22 and zero bytes
+# written to STDOUT.
+stdlib::metadata_get() {
+  local OPTIND opt key outfile
+  local metadata="${METADATA_BASE%/}/computeMetadata/v1"
+  local exit_code
+  while getopts ":k:o:" opt; do
+    case "${opt}" in
+    k) key="${OPTARG}" ;;
+    o) outfile="${OPTARG}" ;;
+    :)
+      return "${E_MISSING_MANDATORY_ARG}"
+      ;;
+    *)
+      stdlib::metadata_get_usage
+      return "${E_UNKNOWN_ARG}"
+      ;;
+    esac
+  done
+  local url="${metadata}/${key#/}"
+
+  stdlib::debug "Getting metadata resource url=${url}"
+  if [[ -z "${outfile:-}" ]]; then
+    curl --location --silent --connect-timeout 1 --fail \
+      -H 'Metadata-Flavor: Google' "$url" 2>/dev/null
+    exit_code=$?
+  else
+    stdlib::cmd curl --location \
+      --silent \
+      --connect-timeout 1 \
+      --fail \
+      --output "${outfile}" \
+      -H 'Metadata-Flavor: Google' \
+      "$url"
+    exit_code=$?
+  fi
+  case "${exit_code}" in
+    22 | 37)
+      stdlib::debug "curl exit_code=${exit_code} for url=${url}" \
+        "(Does not exist)"
+      ;;
+  esac
+  return "${exit_code}"
+}
+
+stdlib::metadata_get_usage() {
+  stdlib::error 'Usage: metadata_get -k <key>'
+  stdlib::info 'For example: metadata_get -k instance/attributes/startup-config'
+}
+
+# Run a command logging the entry and exit.  Intended for system level commands
+# and operational debugging.  Not intended for use with redirection.  This is
+# not named run() because bats uses a run() function.
+stdlib::cmd() {
+  local exit_code argv=("$@")
+  stdlib::debug "BEGIN: stdlib::cmd() command=[${argv[*]}]"
+  "${argv[@]}"
+  exit_code=$?
+  stdlib::debug "END: stdlib::cmd() command=[${argv[*]}] exit_code=${exit_code}"
+  return $exit_code
+}
+
+# Run a command successfully or exit the program with an error.
+stdlib::run_or_die() {
+  if ! stdlib::cmd "$@"; then
+    stdlib::error "run_or_die(): exiting with exit code ${E_RUN_OR_DIE}."
+    exit "${E_RUN_OR_DIE}"
+  fi
+}
+
+stdlib::main() {
+  local OPTIND opt custom_startup_script
+  DELETE_AT_EXIT="$(mktemp -d)"
+  readonly DELETE_AT_EXIT
+  # Set TMPDIR for mktemp() to clean up nicely.
+  declare -xr TMPDIR="$DELETE_AT_EXIT"
```

Comment:
Exporting TMPDIR likely needs to change.  If a daemon spawns in the background and runs after this script exits, the daemon's TMPDIR will be deleted from under it which will likely cause problems.

Reply:
This has been fixed in 9458a1d..ad57d83